docs/howto/use_modules_and_layers.md

```diff
@@ -0,0 +1,270 @@
+}
+```
+
+```scala mdoc:reset:silent
+import zio.ZIO          
+import zio.IO
+import zio.UIO    
+import zio.Has
+import zio.ZEnv
+import zio.ZLayer
+import zio.clock.Clock
+import zio.console.Console
+import zio.random.Random
+
+trait DBError                  
+trait Product
+trait ProductId
+
+case class UserId(value: Long)
+case class User(id: UserId, name: String)
+```
+
+We encountered two new data types `Has` and `ZLayer`, let's get familiar with them.
+
+### The `Has` data type
+
+`Has[A]` represents a dependency on a service of type `A`. Two `Has[_]` can be combined _horizontally_ through `+` and `++` operators, as in
+
+```scala mdoc:invisible
+object Repo {
+  trait Service{}
+}
+
+object Logger {
+  trait Service{
+    def log(s: String): UIO[Unit] = UIO(???)
+  }
+}
+```
+
+```scala mdoc:silent
+val repo: Has[Repo.Service] = Has(new Repo.Service{})
+val logger: Has[Logger.Service] = Has(new Logger.Service{})
+
+val mix: Has[Repo.Service] with Has[Logger.Service] = repo ++ logger
+``` 
+
+At this point you might ask: what's the use of `Has` if the resulting type is just a mix of two traits? Why aren't we just relying on trait mixins? The extra power given by `Has` is that the resulting data structure is backed by an _heterogeneous map_ from service type to service implementation, that collects each instance that is mixed in so that the instances can be accessed/extracted/modified individually, still guaranteeing supreme type safety.
+
+```scala mdoc:silent
+// get back the logger service from the mixed value:
+val log = mix.get[Logger.Service].log("Hello modules!")
+```
+
+As per the recipe above, it is extremely convenient to define a type alias for `Has[Service]`.
+Usually we don't create a `Has` directly, but we do that through `ZLayer`.
+
+### The `ZLayer` data type
+
+`ZLayer[-RIn, +E, +ROut <: Has[_]]` is a recipe to build an environment of type `ROut`, starting from a value `RIn`, possibly producing an error `E` during creation. 
+
+In adherence with environmental concepts, the absence of a required input is represented by `RIn = Any`, conveniently  used in the type alias `ZLayer#NoDeps`.
+
+The `ZLayer` companion object offers a number of constructors to build layers (from pure values, from managed resources to guarantee acquire/release, from other services). 
+
+We can compose `layerA` and `layerB`  _horizontally_ to build a layer that has the requirements of both layers, to provide the capabilities of both layers, through `layerA ++ layerB`
+ 
+We can also compose layers _vertically_, meaning the output of one layer is used as input for the subsequent layer to build the next layer, resulting in one layer with the requirement of the first and the output of the second layer: `layerA >>> layerB` 
+
+### Example: wiring modules together
+Here we define a module to cope with CRUD operations for the `User` domain object. We provide also a live implemntation of the module that depends on a sql connection.
+
+```scala mdoc:silent
```

Comment:
I think some of this should still be available from above?

Reply:
they are defined as simple examples above, but I am actually defining them properly only here